pkg/cloudprovider/providers/azure/auth/azure_auth.go

```diff
@@ -43,29 +43,40 @@ type AzureAuthConfig struct {
 	// The password of the client certificate for an AAD application with RBAC access to talk to Azure RM APIs
 	AADClientCertPassword string `json:"aadClientCertPassword" yaml:"aadClientCertPassword"`
 	// Use managed service identity for the virtual machine to access Azure ARM APIs
-	UseManagedIdentityExtension bool `json:"useManagedIdentityExtension"`
+	UseManagedIdentityExtension bool `json:"useManagedIdentityExtension" yaml:"useManagedIdentityExtension"`
+	// UserAssignedID for explict MSI feature. It only used when UseManagedIdentityExtension is true
+	// If UserAssignedID is empty, it means to use implicit MSI rather than explict MSI
+	UserAssignedID string `json:"userAssignedID" yaml:"userAssignedID"`
 	// The ID of the Azure Subscription that the cluster is deployed in
 	SubscriptionID string `json:"subscriptionId" yaml:"subscriptionId"`
 }
 
 // GetServicePrincipalToken creates a new service principal token based on the configuration
 func GetServicePrincipalToken(config *AzureAuthConfig, env *azure.Environment) (*adal.ServicePrincipalToken, error) {
-	oauthConfig, err := adal.NewOAuthConfig(env.ActiveDirectoryEndpoint, config.TenantID)
-	if err != nil {
-		return nil, fmt.Errorf("creating the OAuth config: %v", err)
-	}
-
 	if config.UseManagedIdentityExtension {
 		glog.V(2).Infoln("azure: using managed identity extension to retrieve access token")
 		msiEndpoint, err := adal.GetMSIVMEndpoint()
 		if err != nil {
 			return nil, fmt.Errorf("Getting the managed service identity endpoint: %v", err)
 		}
+		if len(config.UserAssignedID) > 0 {
+			glog.V(2).Infof("azure: using UserAssingedID (%s) to retrieve access token", config.UserAssignedID)
+			return adal.NewServicePrincipalTokenFromMSIWithUserAssignedID(msiEndpoint,
+				env.ServiceManagementEndpoint,
+				config.UserAssignedID)
+
+		}
+		glog.V(2).Infoln("azure: using implicit MSI to retrieve access token")
```

Comment:
suggest to use `azure: using SystemAssigned MSI to retrieve access token` which is more aligned to official doc.

Reply:
Thx, made the change